README.md
# A7_Dashboard


xml/News
<News>
	<news Title="Dashboard: dxLib: added edit" By="ALw7sH" Date="19/5/2015" />
	<news Title="Dashboard: Main(Tab): added xml files to easily add/remove news,clanwars and members" By="ALw7sH" Date="18/5/2015" />
	<news Title="Dashobard: dxLib: added scrolling ability for gridlists" By="ALw7sH" Date="18/5/2015" />
</News>


dxLib.lua
---------------- Tab
local Tabs = {
Data = {},
Hover = {},
}

function dxCreateTab(text,x,y,w,h,r,g,b)
	local theTab = createElement("dxTab")
	Tabs.Hover[theTab] = false
	Tabs.Data[theTab] = {text=text,x=x,y=y,w=w,h=h,r=r or 255,g=g or 255,b=b or 255,alpha = 100}
	return theTab
end

function dxDrawTab(element,text,x,y,w,h,r,g,b)
	if isElement(element) and getElementType(element) == "dxTab" then
		local data = Tabs.Data[element]
		local text = text or data.text
		local x = x or data.x
		local y = y or data.y
		local w = w or data.w
		local h = h or data.h
		local r = r or data.r
		local g = g or data.g
		local b = b or data.b
		local alpha = data.alpha
		if isMouseInPosition( x, y, w, h ) then
			Tabs.Hover[element] = true
			alpha = math.min(170,alpha+10)
			else
			Tabs.Hover[element] = false
			alpha = math.max(100,alpha-5)
		end
		dxDrawRectangle(x,y,w,h,tocolor(r,g,b,alpha))
		dxDrawText(text,x,y,x+w,y+h,tocolor(r,g,b,255),1,"default-bold","center","center")
		Tabs.Data[element] = {text=text,x=x,y=y,w=w,h=h,r=r,g=g,b=b,alpha = alpha}
	end
end

function dxGetTabText(element)
	if isElement(element) and getElementType(element) == "dxTab" then
		return Tabs.Data[element].text
	end
end

addEventHandler("onClientClick",root,
function(button,state)
	if isDashboardOpen() and button == "left" and state == "down" then
		local tabs = getElementsByType("dxTab")
		if #tabs > 0 then
			for _,tab in ipairs(tabs) do
				local hover = Tabs.Hover[tab]
				if hover == true then
					triggerEvent("onClientdxTabClick",tab,localPlayer)
				end
			end
		end
	end
end
)

---------------- GridList
local GridList = {
Data = {},
Columns = {},
Rows = {},
rAlpha = {}, -- Row alpha
}

function dxCreateGridList(x,y,w,h,searchEdit)
	local theGird = createElement("dxGridList")
	GridList.Data[theGird] = {x=x,y=y,w=w,h=h,searchEdit=searchEdit or nil,hover=false,hoverAlpha=50,scroll=0,maxRows=(math.floor(h/20)-1),rowClicked=0}
	GridList.Columns[theGird] = {}
	GridList.Rows[theGird] = {}
	GridList.rAlpha[theGird] = {}
	return theGird
end

function dxDrawGridList(element)
	if isElement(element) and getElementType(element) == "dxGridList" then
		columns = GridList.Columns[element]
		local rows = GridList.Rows[element]
		local data = GridList.Data[element]
		local x = data.x
		local y = data.y
		local w = data.w
		local h = data.h
		if data.searchEdit then
			local editText = dxEditGetText(data.searchEdit)
			local editText = string.gsub(editText, "([%*%+%?%.%(%)%[%]%{%}%\%/%|%^%$%-])","%%%1")
			if string.len(editText) > 0 then
				local nvm = {}
				for i=1,#rows do
					if string.find(rows[i][1]:lower(),editText:lower()) then
						table.insert(nvm,rows[i])
					end
				end
				rows = nvm
			end
		end
		if isMouseInPosition( x, y, w, h ) then
			GridList.Data[element].hover = true
			GridList.Data[element].hoverAlpha = math.min(80,GridList.Data[element].hoverAlpha+6)
			else
			GridList.Data[element].hover = false
			GridList.Data[element].hoverAlpha = math.max(50,GridList.Data[element].hoverAlpha-3)
		end
		dxDrawRectangle(x,y,w,h,tocolor(255,255,255,GridList.Data[element].hoverAlpha))
		dxDrawRectangle(x,y+25,w,1,tocolor(255,255,255,255))
		if #columns > 0 then
			local posX = x+5
			--- Columns
			for _,column in ipairs(columns) do
				if posX+column.width > x+w then column.width = (x+w) - posX - 5 end
				dxDrawText(column.title,posX,y,posX+column.width,y+25,tocolor(255,255,255,255),1,"default",column.alignX,"center",true)
				posX = posX + column.width
			end
			--- Rows
			if #rows > 0 then
				-- Row Text
				local posX = x+5
				for _=1,#columns do
				local posY = y+30
				if posX+columns[_].width > x+w then columns[_].width = (x+w) - posX - 5 end
					for i=1+data.scroll,math.min(#rows,data.maxRows)+data.scroll do
						if posY+20 < y+h then
							if type(rows[i]) == "table" then
								if type(rows[i][_]) == "string" then
									dxDrawText(rows[i][_],posX,posY,posX+columns[_].width,posY+20,tocolor(255,255,255,255),1,"default",columns[_].alignX,"center",true)
									posY = posY + 20
								end
							end
						end
					end
				posX = posX + columns[_].width
				end
				-- Row Hover
				local posY = y+30
				for i=1+data.scroll,math.min(#rows,data.maxRows)+data.scroll do
				if not GridList.rAlpha[element][i] then GridList.rAlpha[element][i] = 0 end
					if data.rowClicked == i then
						GridList.rAlpha[element][i] = math.min(80,GridList.rAlpha[element][i]+5)
					else
						if isMouseInPosition( x+5 ,posY ,w-10 ,20 ) then
							if getKeyState("mouse1") then
							GridList.Data[element].rowClicked = i
							triggerEvent("onClientGridListRowSelect",element,localPlayer,i)
							end
							GridList.rAlpha[element][i] = math.min(50,GridList.rAlpha[element][i]+5)
						else
							GridList.rAlpha[element][i] = math.max(0,GridList.rAlpha[element][i]-2)
						end
					end
				dxDrawRectangle(x+5,posY,w-10,20,tocolor(255,255,255,GridList.rAlpha[element][i]))
				posY = posY + 20
				end
			end
		end
	end
end

function dxGridListAddColumn(grid,title,width,alignX)
	if isElement(grid) and getElementType(grid) == "dxGridList" then
		table.insert(GridList.Columns[grid],{title=title,width=width,alignX=alignX or "left"})
		return true
	end
end

function dxGridListAddRow(grid,...)
	if isElement(grid) and getElementType(grid) == "dxGridList" then
		table.insert(GridList.Rows[grid],{...})
		return true
	end
end

function dxGridListClear(grid)
	if isElement(grid) and getElementType(grid) == "dxGridList" then
		GridList.Rows[grid] = {}
		return true
	end
end

function dxGridListScroll(state)
	if isDashboardOpen() and state then
		local grids = getElementsByType("dxGridList")
		for i=1,#grids do
			if GridList.Data[grids[i]].hover == true then
				if #GridList.Rows[grids[i]] > GridList.Data[grids[i]].maxRows then
					if state == "mouse_wheel_up" then
						if GridList.Data[grids[i]].scroll > 0 then
							GridList.Data[grids[i]].scroll = GridList.Data[grids[i]].scroll-1
							outputDebugString("Scrolled up")
						end
					elseif state == "mouse_wheel_down" then
						if GridList.Data[grids[i]].scroll < #GridList.Rows[grids[i]]-GridList.Data[grids[i]].maxRows then
							GridList.Data[grids[i]].scroll = GridList.Data[grids[i]].scroll+1
							outputDebugString("Scrolled down")
						end
					end
				end
			end
		end
	end
end
bindKey("mouse_wheel_up","down",dxGridListScroll)
bindKey("mouse_wheel_down","down",dxGridListScroll)

---------------- Edit

local Edit = {
Data = {},
}

function dxCreateEdit(x,y,w,h,defaultText)
	local theEdit = createElement("dxEdit")
	Edit.Data[theEdit] = {x=x,y=y,w=w,h=h,defaultText=defaultText,hoverAlpha=50,Text="",hover=false,clicked=false,selectAll=false}
	return theEdit
end

function dxDrawEdit(element)
	if isElement(element) and getElementType(element) == "dxEdit" then
	local data = Edit.Data[element]
	local x = data.x or 0
	local y = data.y or 0
	local w = data.w or 0
	local h = data.h or 0
	local text = (string.len(data.Text) > 0 and data.Text) or data.defaultText
	if data.clicked == false then
		if isMouseInPosition( x, y, w, h ) then
			Edit.Data[element].hover = true
			Edit.Data[element].hoverAlpha = math.min(80,Edit.Data[element].hoverAlpha+6)
		else
			Edit.Data[element].hover = false
			Edit.Data[element].hoverAlpha = math.max(50,Edit.Data[element].hoverAlpha-3)
		end
	else
		Edit.Data[element].hover = false
		Edit.Data[element].hoverAlpha = math.min(80,Edit.Data[element].hoverAlpha+6)
	end
	dxDrawRectangle(x,y,w,h,tocolor(255,255,255,Edit.Data[element].hoverAlpha))
		if text == data.defaultText then
			dxDrawText(text,x+5,y,x+w-10,y+h,tocolor(200,200,200,200),1,"default","left","center",true)
		else
			dxDrawText(text,x+5,y,x+w-10,y+h,tocolor(255,255,255,255),1,"default","left","center",true)
		end
		if data.selectAll == true then
		local tW = dxGetTextWidth(text)
		dxDrawRectangle(x+5,y,tW,h,tocolor(255,255,255,50))
		end
	dxDrawEmptyRec(x,y,w,h,tocolor(255,255,255,255),1)
	end
end

function dxEditGetText(edit)
	if isElement(edit) and getElementType(edit) == "dxEdit" then
		return Edit.Data[edit].Text
	end
end

addEventHandler("onClientClick",root,
function(button,state)
	if isDashboardOpen() and button == "left" and state == "down" then
		local edit = getElementsByType("dxEdit")
		if #edit > 0 then
			for _,editt in ipairs(edit) do
				local hover = Edit.Data[editt].hover
				if hover == true then
					guiSetInputEnabled(true)
					Edit.Data[editt].clicked = true
					triggerEvent("onClientdxEditClick",editt,localPlayer)
				elseif hover == false then
					guiSetInputEnabled(false)
					Edit.Data[editt].clicked = false
				end
			end
		end
	end
end
)

function addCharacterToEdit(character)
	if isDashboardOpen() then
		local edit = getElementsByType("dxEdit")
		for i=1,#edit do	
			if Edit.Data[edit[i]].clicked == true then
				if Edit.Data[edit[i]].selectAll == true then
					Edit.Data[edit[i]].selectAll = false
					Edit.Data[edit[i]].Text = ""
				end
				Edit.Data[edit[i]].Text = Edit.Data[edit[i]].Text..character
			end
		end
	end
end
addEventHandler("onClientCharacter", getRootElement(), addCharacterToEdit)

function backspace(key,sm)
	if isDashboardOpen() and sm then
		local edit = getElementsByType("dxEdit")
		for i=1,#edit do	
			if Edit.Data[edit[i]].clicked == true then
				if string.len(Edit.Data[edit[i]].Text) > 0 and key == "backspace" then
					if Edit.Data[edit[i]].selectAll == false then
						Edit.Data[edit[i]].Text = string.sub(Edit.Data[edit[i]].Text,1,string.len(Edit.Data[edit[i]].Text)-1)
					else
						Edit.Data[edit[i]].selectAll = false
						Edit.Data[edit[i]].Text = ""
					end
				elseif getKeyState("lctrl") and key == "a" or getKeyState("rctrl") and key == "a" then
					Edit.Data[edit[i]].selectAll = not Edit.Data[edit[i]].selectAll
				end
			end
		end
	end
end
addEventHandler("onClientKey", root, backspace)


util_client.lua
function dxDrawEmptyRec(startX, startY, endX, endY, color, width, postGUI)
	dxDrawLine ( startX, startY, startX+endX, startY, color, width, postGUI )
	dxDrawLine ( startX, startY, startX, startY+endY, color, width, postGUI )
	dxDrawLine ( startX, startY+endY, startX+endX, startY+endY,  color, width, postGUI )
	dxDrawLine ( startX+endX, startY, startX+endX, startY+endY, color, width, postGUI )
end

function dxDrawText_(text,x,y,w,h,color,scale,font,alignX,alignY,clip,wordBreak,postGUI,colorCoded)
	dxDrawText(text,x,y,x + w,y + h,color,scale,font,alignX,alignY,clip or false,wordBreak or false,postGUI or true,colorCoded or true)
end

function isMouseInPosition ( x, y, width, height )
    if ( not isCursorShowing ( ) ) then
        return false
    end
 
    local sx, sy = guiGetScreenSize ( )
    local cx, cy = getCursorPosition ( )
    local cx, cy = ( cx * sx ), ( cy * sy )
    if ( cx >= x and cx <= x + width ) and ( cy >= y and cy <= y + height ) then
        return true
    else
        return false
    end
end

function RGBToHex(red, green, blue, alpha)
	if((red < 0 or red > 255 or green < 0 or green > 255 or blue < 0 or blue > 255) or (alpha and (alpha < 0 or alpha > 255))) then
		return nil
	end
	if(alpha) then
		return string.format("#%.2X%.2X%.2X%.2X", red,green,blue,alpha)
	else
		return string.format("#%.2X%.2X%.2X", red,green,blue)
	end
end

function updateGridlist(grid,thing)
	if isElement(grid) and getElementType(grid) == "dxGridList" then
		if thing == "news" then
			local xml = xmlLoadFile("xml/News.xml")
			for k,v in ipairs(xmlNodeGetChildren(xml)) do
			local title = xmlNodeGetAttribute(v,"Title")
			local by = xmlNodeGetAttribute(v,"By")
			local datee = xmlNodeGetAttribute(v,"Date")
			dxGridListAddRow(grid,title,by,datee)
			end
			xmlUnloadFile(xml)
		elseif thing == "members" then
			local xml = xmlLoadFile("xml/Memebers.xml")
			for k,v in ipairs(xmlNodeGetChildren(xml)) do
			local name = xmlNodeGetAttribute(v,"Name")
			local country = xmlNodeGetAttribute(v,"Country")
			local rank = xmlNodeGetAttribute(v,"Rank")
			dxGridListAddRow(grid,name,country,rank)
			end
			xmlUnloadFile(xml)
		elseif thing == "clanwars" then
			local xml = xmlLoadFile("xml/Clanwars.xml")
			for k,v in ipairs(xmlNodeGetChildren(xml)) do
			local typee = xmlNodeGetAttribute(v,"Type")
			local vs = xmlNodeGetAttribute(v,"Vs")
			local score = xmlNodeGetAttribute(v,"Score")
			local result = xmlNodeGetAttribute(v,"Result")
			dxGridListAddRow(grid,typee,vs,score,result)
			end
			xmlUnloadFile(xml)
		end
	end
end

function getMaps(player)
	if player and isElement(player) then
	triggerServerEvent("Mapshop:clientWantMaps",localPlayer)
	end
end

function setMapsGridList(mapsTable)
	if mapsTable and type(mapsTable) == "table" then
	dxGridListClear(maps)
		for _,map in ipairs(mapsTable) do
			dxGridListAddRow(maps,map)
		end
	end
end
addEvent("Mapshop:sendMapsToClient",true)
addEventHandler("Mapshop:sendMapsToClient",root,setMapsGridList)

addEventHandler("onClientResourceStart",resourceRoot,
function()
getMaps(localPlayer)
end
)


mapshop_server.lua
function getRaceMaps(player)
	local client = player or client
	local mapsTable = {}
	local maps = call(getResourceFromName("mapmanager"), "getMapsCompatibleWithGamemode" , getResourceFromName("race"))
	for k,res in ipairs(maps) do
		table.insert(mapsTable,getResourceInfo(res,"name") or getResourceName(res))
	end
	triggerClientEvent(client,"Mapshop:sendMapsToClient",client,mapsTable)
end
addEvent("Mapshop:clientWantMaps",true)
addEventHandler("Mapshop:clientWantMaps",root,getRaceMaps)

addCommandHandler("refreshMaps",
function (player)
	if isObjectInACLGroup("user."..getAccountName(getPlayerAccount(player)),aclGetGroup("Admin")) then
	for k,v in ipairs(getElementsByType("player")) do
	getRaceMaps(v)
	end
	outputChatBox("[Map shop] #FFFFFFMaps has been refreshed by "..getPlayerName(player),root,0,204,204,true)
	end
end
)


client_main.lua
local sX,sY = guiGetScreenSize()
--local sX,sY = 800,600
Dashboard = {
	Main = {
			rainbow = true,
			bR = 255, bG = 100, bB = 0, -- {bR = 255, bG = 165, bB = 0,},{bR = 0, bG = 204, bB = 204,}
			shader = dxCreateShader("shaders/shader.fx"),
			renderTarget = dxCreateRenderTarget(sX,sY),
			openTick = 0,
			something = -180,
			stats = "close",
			closed = true,
			-- Tabs
			taps = {"Main","Statistic","Map Shop","Garage","Settings"},
			tapsElements = {},
			currentTab = "Main",
			--
			},
	Mapshop = {
			mapInfo = {"Map name","Author","Last time played","Times played","Hunters","Likes"},
			miAlpha = 50,
			},
}

function startDashboard()
	createTabs()
	
	--- Main
		news = dxCreateGridList(20,sY*(80/720),sX-40,sY*(300/720))
			dxGridListAddColumn(news,"Title",sX*(950/1280),"left")
			dxGridListAddColumn(news,"By",sX*(100/1280),"center")
			dxGridListAddColumn(news,"Date",sX*(180/1280),"center")
			updateGridlist(news,"news")
		members = dxCreateGridList(20,sY*((80+300+10)/720),((sX-40)/2)-5,sY*(300/720))
			dxGridListAddColumn(members,"Name",sX*(200/1280),"left")
			dxGridListAddColumn(members,"Country",sX*(200/1280),"center")
			dxGridListAddColumn(members,"Rank",sX*(200/1280),"center")
			updateGridlist(members,"members")
		clanwars = dxCreateGridList((sX-40)-((sX-40)/2)+25,sY*((80+300+10)/720),((sX-40)/2)-5,sY*(300/720))
			dxGridListAddColumn(clanwars,"Type",sX*(100/1280),"left")
			dxGridListAddColumn(clanwars,"VS",sX*(200/1280),"left")
			dxGridListAddColumn(clanwars,"Score",sX*(100/1280),"center")
			dxGridListAddColumn(clanwars,"Result",sX*(200/1280),"center")
			updateGridlist(clanwars,"clanwars")
	--
	--- Map shop
		mapsEdit = dxCreateEdit(20,(sY*((80+300+10)/720)),sX-40,25,"Map name...")
		maps = dxCreateGridList(20,sY*(80/720),sX-40,sY*(300/720),mapsEdit)
			dxGridListAddColumn(maps,"Map name",sX*(501/1280),"left")
	--
end
addEventHandler("onClientResourceStart",resourceRoot,startDashboard)

function stopDashboard()

end
addEventHandler("onClientResourceStop",resourceRoot,startDashboard)

function dashboardRender()
	if Dashboard.Main.rainbow == true then
	
	end
	dxSetRenderTarget(Dashboard.Main.renderTarget)
	dxDrawRectangle(0,0,sX,sY,tocolor(Dashboard.Main.bR,Dashboard.Main.bG,Dashboard.Main.bB,255))
	dxDrawRectangle(20,sY*(50/720),sX-40,2,tocolor(255,255,255,255))
	dxDrawRectangle(sX-20-75,sY*(50/720)-25,75,25)
	dxSetBlendMode("modulate_add")
	dxDrawText("A7D 1.0.0",sX-20-75,sY*(50/720)-25,(sX-20-75)+75,(sY*(50/720)-25)+25,tocolor(Dashboard.Main.bR,Dashboard.Main.bG,Dashboard.Main.bB,255),1.2,"default","center","center")
	dxSetBlendMode("blend")
	drawTabs()
	
	if Dashboard.Main.currentTab == "Main" then
		dxDrawGridList(news)
		dxDrawGridList(members)
		dxDrawGridList(clanwars)
	elseif Dashboard.Main.currentTab == "Map Shop" then
		dxDrawGridList(maps)
		dxDrawEdit(mapsEdit)
			if isMouseInPosition( (sX/2)-(500/2),sY*((80+300+10+10)/720)+25,sX*(500/1280),sY*((300-10)/720)-25 ) then
				Dashboard.Mapshop.miAlpha = math.min(80,Dashboard.Mapshop.miAlpha+6)
			else
				Dashboard.Mapshop.miAlpha = math.max(50,Dashboard.Mapshop.miAlpha-3)
			end
		dxDrawRectangle((sX/2)-(500/2),sY*((80+300+10+10)/720)+25,sX*(500/1280),sY*((300-10)/720)-25,tocolor(255,255,255,Dashboard.Mapshop.miAlpha))
		local sizeY = (sY*((300-10)/720)-25)/#Dashboard.Mapshop.mapInfo
		local posY = sY*((80+300+10+10)/720)+25
		for i=1,#Dashboard.Mapshop.mapInfo do
			local tH = dxGetTextWidth(Dashboard.Mapshop.mapInfo[i]..":")
			dxDrawText(Dashboard.Mapshop.mapInfo[i]..":",(sX/2)-((sX*(500/1280))/2)+5,posY,(sX/2)-((sX*(500/1280))/2)+(tH),posY+sizeY,tocolor(255,255,255,255),1,"default","left","center",false)
			dxDrawText(" ".."ALw7sH!",(sX/2)-((sX*(500/1280))/2)+5+tH,posY,(sX/2)-((sX*(500/1280))/2)+(sX*(500/1280)),posY+sizeY,tocolor(255,255,255,255),1,"default","left","center",true)
			posY = posY + sizeY
		end
	end
	dxSetRenderTarget()
	
	if Dashboard.Main.stats == "open" then
		local tick = getTickCount() - Dashboard.Main.openTick
		local progress = math.min(tick/1000,1)
		rot = interpolateBetween(Dashboard.Main.something,0,0,0,0,0,progress,"OutQuad")
		Dashboard.Main.something = rot
	elseif Dashboard.Main.stats == "close" then
		local tick = getTickCount() - Dashboard.Main.openTick
		local progress = math.min(tick/1000,1)
		rot = interpolateBetween(Dashboard.Main.something,0,0,-180,0,0,progress,"OutQuad")
		Dashboard.Main.something = rot
		if progress >= 1 then
			outputDebugString("Dashboard: Closed")
			Dashboard.Main.closed = true
			removeEventHandler("onClientRender",root,dashboardRender)
		end
	end
	
	dxSetShaderTransform(Dashboard.Main.shader,0,rot,0,0,1,0,false,0,0,false)
	dxSetShaderValue(Dashboard.Main.shader,"A7D",Dashboard.Main.renderTarget)
	dxSetBlendMode("modulate_add")
	dxDrawImage(0,0,sX,sY,Dashboard.Main.shader)
	dxSetBlendMode("blend")
end

function createTabs()
local posX = 20
	for _,text in ipairs(Dashboard.Main.taps) do
	local tab = dxCreateTab(text,posX,sY*(50/720)-25,75,25)
	table.insert(Dashboard.Main.tapsElements,tab)
	posX = posX + 75
	end
end

function drawTabs()
	for k,v in ipairs(Dashboard.Main.tapsElements) do
	dxDrawTab(v)
	end
end

addEvent("onClientdxTabClick")
addEventHandler("onClientdxTabClick",root,
function()
	Dashboard.Main.currentTab = dxGetTabText(source)
end
)

function toggleDashboard()
	if Dashboard.Main.stats == "close" then
		if Dashboard.Main.closed == true then
			addEventHandler("onClientRender",root,dashboardRender)
			Dashboard.Main.closed = false
		end
		Dashboard.Main.openTick = getTickCount()
		Dashboard.Main.stats = "open"
		showCursor(true)
		showChat(false)
	else
		Dashboard.Main.openTick = getTickCount()
		Dashboard.Main.stats = "close"
		showCursor(false)
		showChat(true)
	end
end
bindKey("F7","down",function() toggleDashboard() end)

function isDashboardOpen()
	if Dashboard.Main.stats == "open" then
	return true
	else
	return false
	end
end
